### PR TITLE
Mainline geo t

### PR DESCRIPTION
**Pull request by coreyjadams**
# PhysicsNeMo Pull Request

This PR moves GeoTransolver out of experimental and into physicsnemo proper.  Because GeoT also has a FLARE attention backend, we also move FLARE at the same time.

A few other pieces change with this:
- FLARE now has full TransformerEngine support.
- GeoTransolver's OOD component is transformed from an integrated-directly piece of the model to a wraps-the-model component.
- GALE, flare_attention move to physicsnemo.nn
- Context_projector stays with GeoTransolver in the models folder.

Tests are updated, and most examples are changed only to update import paths, etc.  The major update is the OODGuard changes.  No example yet is exercising the FLARE TE path.

A docs update will come in a follow on PR so it can be reviewed on a separate time scale.

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

**Review by peterdsharpe — APPROVED** [review]
Verified changes on the unified recipe (CODEOWNER), LGTM there!

One question - where is the new `geotransolver_guard` coming from? Is that splitting out functionality that used to be in core `geotransolver.py`?

**Comment by CharlelieLrt** [comment]
This PR is breaking [the kinetic Monte-Carlo recipe](https://github.com/NVIDIA/physicsnemo/blob/465387f5774de8d47717a4daa5a79df0a965466a/examples/kinetic_monte_carlo/utils/nn.py#L30-L31) as well as checkpoints that were distributed. Nothing against the move, but can we keep it importable from `experimental` at least for some time (maybe with warning or something)?

**Comment by coreyjadams** [comment]
>One question - where is the new geotransolver_guard coming from? Is that splitting out functionality that used to be in core geotransolver.py?

Yes, exactly.  This is really something that @mnabian needs to weigh in on, if we should build it this way or another way.

**Comment by CharlelieLrt** [comment]
Also, model hould be added to API doc I think?

**Comment by coreyjadams** [comment]
> Also, model hould be added to API doc I think?

I would like to put in a second docs PR to give @megnvidia breathing room to review independent of code changes.  I've got the updates staged locally already.  Are you ok with that?  We can have it ready up here before approval of this one.

**Comment by coreyjadams** [comment]
/ok to test [8a20c73](https://github.com/NVIDIA/physicsnemo/pull/1845/commits/8a20c73598a97c715ce1e0fcfc9fd9bb4c00c7f3)

File: physicsnemo/experimental/models/geotransolver/__init__.py [hunk 1]
```diff
@@ -73,24 +73,23 @@
...
     ContextProjector,
     GeometricFeatureProcessor,
     GlobalContextBuilder,
     MultiScaleFeatureExtractor,
     StructuredContextProjector,
 )
-from .gale import (
-    GALE,
-    GALE_FA,
-    GALE_block,
```
Comment (CharlelieLrt) [inline physicsnemo/experimental/models/geotransolver/__init__.py:90]: nit: `GALEBlock`

**Comment by coreyjadams** [comment]
Docs update in #1847 .  It's this whole PR + docs, so this one needs to merge first.

**Comment by coreyjadams** [comment]
@CharlelieLrt I've added shims back to experimental.  @ktangsali maybe a good test is to try this PR in physicsnemo-cfd and validate the checkpoints import correctly?  I can do it if you think that will be robust testing.

**Review by CharlelieLrt — CHANGES_REQUESTED** [review]
Kinetic Monte-Carlo recipe to update with right import

File: examples/cfd/external_aerodynamics/transformer_models/README.md [hunk 2]
```diff
@@ -43,7 +43,7 @@ To train the model, first we compute normalization factors on the dataset to mak
 
 > By default, the normalization sets the mean to 0.0 and std to 1.0 of all labels in the dataset, computing the mean across the train dataset. You could adapt this to a different normalization, however take care to update both the preprocessing as well as inference scripts. Min/Max is another popular strategy.
 
-To configure your training run, use `hydra`. The config contains sections for the model, data, optimizer, and training settings. For details on the model parameters, see the API for `physicsnemo.models.transolver` and `physicsnemo.experimental.models.geotransolver`.
+To configure your training run, use `hydra`. The config contains sections for the model, data, optimizer, and training settings. For details on the model parameters, see the API for `physicsnemo.models.transolver` and `physicsnemo.models.geotransolver`.
```
Comment (megnvidia) [inline examples/cfd/external_aerodynamics/transformer_models/README.md]:
```suggestion
To configure your training run, use `hydra`. The config contains sections for the model, data, optimizer, and training settings. For details on the model parameters, refer to the API for `physicsnemo.models.transolver` and `physicsnemo.models.geotransolver`.
```

File: examples/structural_mechanics/crash/README.md [hunk 3]
```diff
@@ -213,26 +213,38 @@ Predicted meshes are written as .vtp files under `./predicted_vtps/`, and can be
...
 
 - **Global parameters** — per-channel bounding box on the global embedding
   (e.g. `velocity_x`, `thickness_scale`).
 - **Geometry** — k-nearest-neighbour distance on a pooled geometry latent.
 
-Enable it through the model config by setting `guard_config` to a mapping
-(leave `null` to disable):
+To enable the guard, wrap a constructed `GeoTransolver` with
+`GuardedGeoTransolver` (from `physicsnemo.experimental.guardrails.embedded`).
+The wrapper observes the two surfaces above via a forward hook and delegates the
```
Comment (megnvidia) [inline examples/structural_mechanics/crash/README.md]:
```suggestion
The wrapper observes the two surfaces above through a forward hook and delegates the
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -10,6 +10,15 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
+- Promotes GeoTransolver out of `experimental` to
+  `physicsnemo.models.geotransolver.GeoTransolver`, together with the FLARE
+  model (`physicsnemo.models.flare.FLARE`) and the reusable GALE and FLARE
+  attention layers (`physicsnemo.nn.GALE`, `physicsnemo.nn.GALE_block`,
+  `physicsnemo.nn.FLARE`). The embedded OOD guard is decoupled from the model:
```
Comment (megnvidia) [inline CHANGELOG.md]:
```suggestion
  `physicsnemo.nn.FLARE`). The embedded OOD guard is decoupled from the model.
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -10,6 +10,15 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+  wrap a constructed GeoTransolver with
```
Comment (megnvidia) [inline CHANGELOG.md]:
```suggestion
Wrap a constructed GeoTransolver with 
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -10,6 +10,15 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+  `physicsnemo.experimental.guardrails.embedded.GuardedGeoTransolver` (or
```
Comment (megnvidia) [inline CHANGELOG.md]:
```suggestion
  GuardedGeoTransolver (or call 
```

File: CHANGELOG.md [hunk 7]
```diff
@@ -10,6 +10,15 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+  `attach_ood_guard`) to enable out-of-distribution guarding — the
```
Comment (megnvidia) [inline CHANGELOG.md]:
```suggestion
  `attach_ood_guard`) to enable out-of-distribution guarding. The `guard_config` 
```

File: CHANGELOG.md [hunk 8]
```diff
@@ -10,6 +10,15 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+  `guard_config` model argument is removed.
```
Comment (megnvidia) [inline CHANGELOG.md]:
```suggestion
  model argument is removed.
```

**Review by megnvidia — APPROVED** [review]

File: test/models/geotransolver/test_geotransolver.py [hunk 9]
```diff
@@ -14,11 +14,10 @@
 # See the License for the specific language governing permissions and
```
Comment (mnabian) [inline test/models/geotransolver/test_geotransolver.py:14]: You can also include a test for the `return_embedding_states=True` path.

File: physicsnemo/nn/module/flare_attention.py [hunk 10]
```diff
@@ -144,14 +187,17 @@ def __init__(
...
         self.self_v = linear_layer(dim_head, dim_head)
 
-        # te attention
+        # Transformer Engine cross-attention supports the unequal global and
+        # token sequence lengths used by both FLARE attention passes.
         if self.use_te:
             self.attn_fn = te.DotProductAttention(
                 num_attention_heads=self.heads,
                 kv_channels=self.dim_head,
                 attention_dropout=dropout,
```
Comment (mnabian) [inline physicsnemo/nn/module/flare_attention.py]: When use_te=True, attention-weight dropout is applied inside the TE kernel in addition to the shared out_dropout, but the non-TE path skips attention dropout entirely. This won't affect our GeoTransolver trainings as we usually don't use dropout, but will affect the models that use Concrete Dropout for UQ, and makes the two modes non-equivalent for UQ.

**Review by mnabian — APPROVED** [review]
LGTM! Left a couple of minor comments. Great to see TE is now enabled for FLARE! Guardrail changes look good, but an action item for me is to update the user documentation accordingly.

File: test/models/geotransolver/test_geotransolver.py (hunk 9 above)
Comment (coreyjadams) [reply test/models/geotransolver/test_geotransolver.py:14]: Done.

File: physicsnemo/nn/module/flare_attention.py (hunk 10 above)
Comment (coreyjadams) [reply physicsnemo/nn/module/flare_attention.py]: Good catch here.

I've update this: TE attention dropout is now 0.0, so dropout remains outside the attention layer.  I think this should let us apply concrete dropout correctly too, even when using TE too.

**Review by CharlelieLrt — APPROVED** [review]
LGTM for the changes I requested

**Review by loliverhennigh — APPROVED** [review]
Rubber Stamp

**Review by ktangsali — APPROVED** [review]
Changes to TOML look good

File: physicsnemo/experimental/guardrails/embedded/geotransolver_guard.py [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+    """
+    if "global_embedding" in kwargs:
+        return kwargs["global_embedding"]
+    if len(args) >= 3:
+        return args[2]
+    return None
+
+
+class GuardedGeoTransolver(nn.Module):
+    """GeoTransolver wrapped with an out-of-distribution guard.
```
[the remark this answers is not in the document]
Comment (coreyjadams) [reply physicsnemo/experimental/guardrails/embedded/geotransolver_guard.py]: Fixed.

File: physicsnemo/experimental/guardrails/embedded/geotransolver_guard.py [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+    tokenizer = getattr(model.context_builder, "global_tokenizer", None)
+    if tokenizer is None:
+        return None
+    return int(tokenizer.in_project_x.in_features)
+
+
+def _extract_global_embedding(args: tuple, kwargs: dict) -> torch.Tensor | None:
+    """Pull ``global_embedding`` from a GeoTransolver forward call.
+
+    ``GeoTransolver.forward`` takes ``global_embedding`` as its third positional
```
[the remark this answers is not in the document]
Comment (coreyjadams) [reply physicsnemo/experimental/guardrails/embedded/geotransolver_guard.py]: Fixed.

**Comment by coreyjadams** [comment]
/ok to test [2bea253](https://github.com/NVIDIA/physicsnemo/pull/1845/commits/2bea253e3e763f92e69f8f625387174d433f30bf)

**Comment by coreyjadams** [comment]
/ok to test [670c7a4](https://github.com/NVIDIA/physicsnemo/pull/1845/commits/670c7a43af0d82bb04381adae8975723e21cff21)

**Comment by coreyjadams** [comment]
/ok to test [486cd6c](https://github.com/NVIDIA/physicsnemo/pull/1845/commits/486cd6c5323addd49d6a8d8680d06ef53f16d8f0)
